### PR TITLE
Adds ADR for remapping PublicationInformation

### PR DESCRIPTION
**Pull request by JPrevost**
also adds config to allow adr cli to create an adr in the same directory other adrs are found in this repository

https://mitlibraries.atlassian.net/browse/GDT-199

Rendered markdown of current ADR: https://github.com/MITLibraries/transmogrifier/blob/gdt-199-publication-adr/docs/adrs/0003-support-aggregations-on-publisher-name.md

### Includes new or updated dependencies?
NO

### Changes expectations for external applications?
YES

### CODE REVIEW

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (ghukill) [inline docs/adrs/0003-support-aggregations-on-publisher-name.md:49]: From EngX's / API POV, what makes a single value simpler than an array?  Is it parsing those results from the actual document in GraphQL?

**Comment by ghukill** [comment]
At this time, I'm leaning towards Option 2 for a couple of reasons.

First, I found this quite compelling:
> This decouples our data model more from MARC where we seem to have modeled a set of data to match 260/264 in a way that doesn't seem to map to our other sources

Second, when you extend that thinking further, though I'm confident we _will_ encounter publisher dates, it's more likely that we'd want those dates when already in the context (i.e. facet filter) of thinking about dates.  This, as opposed to rooting around in an object for them.

This might suggest a somewhat general pattern where _some_ fields in TIMDEX benefit from having `kind` or other kinds of qualifiers, e.g. Dates and Locations.  We may have date and location information for other descriptive components -- like Publisher -- but we often don't need to know all that at once.  Storing a `date@kind=publisher` and `location@kind=publisher`, as long as the `kind` values are sensible and controlled (which we could enforce in Transmog), makes that information more where you need and expect it.

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (JPrevost) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:49]: It's less EngX than end users I'm thinking about. Nested objects can be more difficult to work with in GraphQL than the equivalent top level field. And strings are easier than arrays to work with initially. This doesn't mean we should aim for totally flat records of entirely strings, just that when our data does not require objects or arrays, we should lean towards strings.

Comment (jazairi) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:51]: It seems like we should make it an array given that it's possible for multiple occurrences, but I agree that it would be useful to find out how often that happens, if ever, to inform that decision. It would be kind of silly if we accommodated repeatable MARC fields that are never actually repeated.

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (jonavellecuerdo) [inline docs/adrs/0003-support-aggregations-on-publisher-name.md:71]: @JPrevost For Option 2, would it also require only choosing the _first_ occurrence of the MARC 260 and 264 fields? 🤔

Comment (JPrevost) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:71]: Good question! I believe similar to the discussion in Option 1, the fields are repeatable but it's unclear how frequently they are actually repeatable. I'd suggest we may want to look at our data to decide if this should be an Array of Strings or a single String. I suspect an Array of Strings will be where we end up but I'd hate to make that decision based on suspicion rather than data.

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (ghukill) [inline docs/adrs/0003-support-aggregations-on-publisher-name.md:75]: Curious about this line:
> All sources except Alma remap to use `Publisher` instead of `PublicationInformation` with no other changes at this time.

I realized on my first read, I had made the assumption:
- all sources (including Alma) would begin to write the publisher *name* to this new multivalued string `publisher` field
- future work: any sources (including Alma) that have publisher date or location information, could write to other appropriate fields, e.g. `dates` or `locations` with a qualifier like `@kind="Publisher"`

Does this align with your thinking and the wording here?

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (ghukill) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:71]: Just to tease this out a bit more... if:

- `publisher` is a multivalued string field
- and any other publisher information gets written to other fields like `dates` or `locations` ([see comment below](https://github.com/MITLibraries/transmogrifier/pull/132/files#r1506065717))

wouldn't this mean, for Option 2, that we could accomodate both MARC 260 and 264?  Stated more generally, that Option 2 supports multiple publishers by virtue of the new `publisher` field being multivalued and all other values mapped to other, multivalued fields?

Comment (JPrevost) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:71]: Correct. If we just make `publisher` an array of Strings, we can support any source that currently or in the future has multiple values for all aspects of this (dates, names, locations).

If we make `publisher` a single string, we could still support multiple Dates/Locations if that is the only thing that is truly multivalued in our sources. That may be a bit hard to tease out if that is all we truly need or if we should just accept that supporting multivalued is better because we just aren't sure.

**Comment by ghukill** [comment]
> At this time, I'm leaning towards Option 2 for a couple of reasons.
> 
> First, I found this quite compelling:
> 
> > This decouples our data model more from MARC where we seem to have modeled a set of data to match 260/264 in a way that doesn't seem to map to our other sources
> 
> Second, when you extend that thinking further, though I'm confident we _will_ encounter publisher dates, it's more likely that we'd want those dates when already in the context (i.e. facet filter) of thinking about dates. This, as opposed to rooting around in an object for them.
> 
> This might suggest a somewhat general pattern where _some_ fields in TIMDEX benefit from having `kind` or other kinds of qualifiers, e.g. Dates and Locations. We may have date and location information for other descriptive components -- like Publisher -- but we often don't need to know all that at once. Storing a `date@kind=publisher` and `location@kind=publisher`, as long as the `kind` values are sensible and controlled (which we could enforce in Transmog), makes that information more where you need and expect it.

Further support of Option 2, seems like there is similar thinking in MODS for `<origininfo>` where the publisher + event is recorded, but then associated dates (and could conceptually extend this to locations) are stored in that respective part of the record: https://www.loc.gov/standards/mods/userguide/origininfo.html.

**Review by ehanson8 — COMMENTED** [review]
Agree with everyone else so far on Option 2, the decoupling from MARC will make the MARC transform logic more complicated but I think it's better in the long run

**Comment by ghukill** [comment]
FWIW, this might suggest that Option 1 (array of complex publisher objects) aligns more closely with MODS: https://www.loc.gov/standards/mods/userguide/origininfo.html.

In the MODS schema, the `<origininfo>` parent element has child elements for place, publisher, date, location, etc. 

To me, the difference between Option 1 and 2 feels like it gets at a bigger data model decision: do we store subjects/dates/locations under complex objects like publishers, contributors, etc.?  or do we store those as simple arrays of strings, and then qualify any subects/dates/locations with a `@kind` qualifier and store them in `subjects`/`dates`/`locations`/etc.?

The decision for this ADR, feels like it would set precedence for that going forward, as I'm not seeing this situation in any other fields (complex objects with child dates/locations).  

I think that I continue to support Option 2, and lean into that model:
- simple, top level arrays of strings where it makes sense
- any subjects / dates / locations get routed to those fields, with meaningful and ideally somewhat controlled `@kind` qualifiers

To me, this is more flexibile in the long run.  As we encounter subjects / dates / locations that should be in the record, but don't fall under a complex object, we can continue to slot them into those fields, but ensure (and document) what the `@kind` qualifiers are.

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (jonavellecuerdo) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:75]: @JPrevost Does this change mean `TimdexRecord.publication_information`/`PublicationInformation` will be deprecated? 🤔

**Comment by jonavellecuerdo** [comment]
**Option 2** sounds good to me, especially after looking into `all-current` index on OpenSearch (shared in [this comment](https://github.com/MITLibraries/transmogrifier/pull/132#discussion_r1507657483). Approved on my end once decision is officially written into the ADR!

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (JPrevost) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:71]: Awesome, thanks for the peek into the data. This is extremely helpful context. I feel like I have enough info to add some additional context to the Options and propose a Decision (option 2 seems popular with the top level array for publisher names and everything else broken into Location and Dates).

**Comment by jazairi** [comment]
In case any further coalescence is needed, I support Option 2 at this point, if not just for flexibility and ease of use.

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (matt-bernhardt) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:51]: What sort of warnings get generated when Transmog encounters multiple values when a single one is expected? I'm trying to understand what would happen if, based on an analysis of our current catalog, we decide that multiple values are not supported - but next month something changes and we start seeing them.

**Review by matt-bernhardt — COMMENTED** [review]
For those cases where we have multiple publishers (not just multiple publication years), would Option 2 introduce potential confusion about the relationship between years, locations, and names among the publishers? Am I interpreting these options correctly with these arrangements?

## Option 1
```
{
  "publicationInfo": [
    {
      "name": "Massachusetts Institute of Technology",
      "location": "Cambridge, MA",
      "date": "2014"
    },
    {
      "name": "Wiley",
      "location": "Hoboken, NJ",
      "date": "2016"
    }
  ]
}
```

## Option 2
```
{
  "date": [
    {
      "kind": "Publication Date",
      "value": "2016"
    },
    {
      "kind": "Publication Date",
      "value": "2014"
    }
  ],
  "location": [
    {
      "kind": "Publisher Location",
      "value": "Cambridge, MA"
    },
    {
      "kind": "Publisher Location",
      "value": "Hoboken, NJ"
    }
  ],
  "publisher": [
    "Massachusetts Institute of Technology",
    "Wiley"
  ]
}
```

I'm not sure how much of an issue this is - it might be fine.

**Comment by JPrevost** [comment]
@matt-bernhardt great question. Yes, Option 1 provides an ability Option 2 does not that we hadn't stated (and I hadn't noticed). Namely, if there is value in keeping the publisher related date together when there are multiple publishers, Option 1 provides that in a way that Option 2 does not.

Based on how infrequently there appear to be multiple _real_ publishers (versus currently a lot of the extra publishers are actually just years inserted into the array), this may not be a concern. However, it should be noted as a distinct feature difference between the two options for sure and noted in the consequences section if we choose the one that would lose this ability to keep these fields together in an object.

**Comment by ghukill** [comment]
Agreed, that @matt-bernhardt's observation is an important one, namely that we decouple the publisher name from the its dates/locations and lose that tether.

Maybe this is not problematic if we continue to view these TIMDEX records through a discovery lens, and not a canonical record lens.  And, potentially even _more_ accurate when we consider the behavior of facets.

Consider this situation with two publishers:

1. "Great Writings; Bend, OR; 1980"
2. "Amazon Reprints, Seattle, WA; 2020"
3. "Ebooks Inc., Believed to be 2023 but records are hazy"

If this record comes up in a search, and a user clicks the publisher facet for `"Great Writings"`, if we tightly couple the publisher name + location + date, we'd _lose_ the location "Seattle, WA" and the date "2020" in the facets.  However, I'd float those facet values are still relevant to the record, as they are part of its metadata in its totality.

The flipside -- and what I assume is the concern raised above -- is that by decoupling them, we would lose the ability to look directly at this TIMDEX record and know:

1. For the "Great Writings" publisher, the date was 1980 and the location was Bend, OR
2. For the "Amazon Reprints" publisher, the date was 2020 and the location was Seattle, WA

I hesitate to suggest it... but what if we combined Option 1 & 2?

```json
{
  "publishers": [
    {
      "name": "Great Writings",
      "date": "1930",   <----- use original string
      "location": "Bend, OR"
    },
    {
      "name": "Amazon Reprints",
      "date": "2020",   <----- use original string
      "location": "Seattle, WA"
    },
    {
      "name": "Ebooks Inc.",
      "date": "Believed to be 2023 but records are hazy"   <----- use original string
    }
  ],
  "date": [
    {
      "kind": "Published",
      "value": "1980-01-01"   <----- parsed as true dates, with an eye towards this desired state
    },
    {
      "kind": "Published",
      "value": "2020-01-01"   <----- parsed as true dates, with an eye towards this desired state
    }
  ],
  "locations": [
    {
      "kind": "Published",
      "value": "Bend, OR"
    },
    {
      "kind": "Published",
      "value": "Seattle, WA"
    }
  ]
}
```

In this scenario, we'd get:

- accurate facets; no loss of information when one publisher identified
- publisher dates/locations slot into `dates` and `locations` for ease of use there
  - furthermore, they are parsed and normalized according to those fields
- for the messy "Ebooks Inc." entry, we maintain the original strings, but we just skip extracting and normalizing dates from it

There is duplication of information, but we'ev also normalized and filtered out bad values while still retaining the original which may have some meaning.

It also might be less jarring to see this in the context of other fields.  If the pattern is: store complex objects, but attempt to extract meaningful dates / subjects / locations from it and store in those respective fields, then maybe it doesn't seem as unusual?

Lastly, from a Transmogrifier POV, we could focus on writing complex Publisher objects, then have shared methods and normalization logic across all sources that attempt to parse out distinct subjects, dates, locations, etc.

Just food for thought!

**Comment by ehanson8** [comment]
> Maybe this is not problematic if we continue to view these TIMDEX records through a discovery lens, and not a canonical record lens.

I still think discovery is the most appropriate lens to view TIMDEX and combining 1 & 2 feels like it's introducing a lot of complexity for accuracy in a likely rare occurrence of multiple legitimate publishers

**Comment by ghukill** [comment]
More comments forthcoming, but I did want to correct a mistake in my [comment from last week](https://github.com/MITLibraries/transmogrifier/pull/132#issuecomment-1973710461).

Specifically, that when selecting a facet value, we don't "lose" the other facet values from that particular record.  See this example of selecting "Polygon data" as a facet value, but because the document contains other values for "Content Type", they still show up:

Screenshot 2024-03-04 at 10 22 44 AM

More comments coming, but this is an important clarification as neither Option 1 or 2 result in the behavior I had suggested above.

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (ghukill) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md]: @JPrevost - language updated in Option 2, per [your aggrement with a slight rewording](https://github.com/MITLibraries/transmogrifier/pull/132#discussion_r1506082129), per this commit: https://github.com/MITLibraries/transmogrifier/pull/132/commits/1dfcfab19a41d5a2c021658e7acd9186c79afb53.

**Comment by ghukill** [comment]
@JPrevost, @matt-bernhardt, @ehanson8, @jazairi, @jonavellecuerdo - an "Option 3" has been added [per this commit](https://github.com/MITLibraries/transmogrifier/pull/132/commits/99173c3fcceafb7b6daca126d0507c333a7970e4) that sketches a combination of Option 1 and Option 2.

Matt's [observation](https://github.com/MITLibraries/transmogrifier/pull/132#pullrequestreview-1911456231) about Option 2 losing context between publisher names and dates/locations felt like an important one.  Personally, this felt like a deal breaker for Option 2 for me.

However, I feel Option 1 may continue to propagate some pain points we are feeling in the TimdexRecord / Opensearch / GraphQL / Search UI / Item Page UI relationshipos where we are required to dig into complex objects for aggregations.  This alone may not be technically problematic, but it could be brittle in the longer term (see example in Option 3 where "Publisher" or "Date" sections start to drift between the search UI filters and the item page representation).

It's hard to succinctly summarize, and might be beyond the scope of this ADR to some degree, but to me Option 3 represents a pattern that could be applied going forward: **store complex objects to preserve context, but also extract subjects / dates / locations for ease of use and consistency of aggregation and item viewing**.

Perhaps Option 3 is not ultimately what we want, but hoping it can help clarify the strenghts and weaknesses of the current Options 1 and 2.

At this time, my vote would be for Option 3.

**Comment by jonavellecuerdo** [comment]
I like **Option 3**. I think [lines 165-168 of the ADR](https://github.com/MITLibraries/transmogrifier/blob/gdt-199-publication-adr/docs/adrs/0003-support-aggregations-on-publisher-name.md?plain=1#L165-L168) summarize it really well. I feel it benefits the end user (by preventing information loss) and developers (by easing navigation of the data model). 

As a side note, [`Aardvark.dct_issue_s `could be updated to use `kind="Publication date"` (instead of "Issued")](https://github.com/MITLibraries/transmogrifier/blob/0fd8ea882733367606f8c13610b3d000a94beef8/transmogrifier/sources/json/aardvark.py#L222-L225) to follow the `timdex.Date` derivations for other transformers.

**Comment by ghukill** [comment]
> As a side note, [`Aardvark.dct_issue_s `could be updated to use `kind="Publication date"` (instead of "Issued")](https://github.com/MITLibraries/transmogrifier/blob/0fd8ea882733367606f8c13610b3d000a94beef8/transmogrifier/sources/json/aardvark.py#L222-L225) to follow the `timdex.Date` derivations for other transformers.

Agreed @jonavellecuerdo about looking for opportunities to standardize the `kind` values we use.  To the degree that we "extract" data from other parts of a record and store them as Dates / Subjects / Locations, it increases the value in standardizing the `kind` qualifiers we use.

It may be out of scope for implementing the final decision of this particular ADR, but a follow-up ADR at some point around normalizing `kind` values would be a great one.

**Comment by ehanson8** [comment]
I'm fine with Option 3 as well. Standardizing `kind` values is something we've wanted to do but just never got around to it, this would be good to prioritize as we go down this road

**Review by jonavellecuerdo — APPROVED** [review]

**Review by jazairi — APPROVED** [review]
Option 3 sounds good to me, too. I agree with @ghukill that, from the perspective of the API consumer, this structure is no more complex to navigate than the one presented in Option 2. That it also provides richer and clearer data makes this an easy choice for me.

Thanks, all, for the nuanced and thoughtful discussion on this, leading to another excellent decision. 🙂

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (ehanson8) [inline docs/adrs/0003-support-aggregations-on-publisher-name.md:176]: Can this be `PublicationInformation`? We aren't abbreviating any of the other `transmogrifier` field names or classes

**Review by ehanson8 — APPROVED** [review]
Approved but one requested name change

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (ghukill) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:176]: I'm a little fuzzy on GraphQL deprecation, but perhaps that is contributing to this variation on `PublicationInformation` which already existed?

Is it correct to assume @JPrevost, @jazairi, @matt-bernhardt, that if we changed the field type of the **pre-existing** `PublicationInformation` field, that doesn't give us the ability to really deprecate it?  Because from GraphQL's POV the field name didn't change, so no ability to say, "This field is deprecated, use XYZ..."?

If so, @ehanson8 and @jazairi (as thumbs-uppers for this suggestion), what would you think about `Publishers` as the complex object field name?  My assumption is that "Information" or "Info" was originally proposed in the event we don't have a name, where just a date or location feels a little odd as a "Publisher".  But, that's generally only the case for copyright dates, and we could elect to skip a `Publisher` entry if only that copyright date, and instead just write that directly to `Dates`.

All considered, I suppose my vote would be `Publishers` as the top level field for complex objects, thereby matching other field name patterns like `Contributors`.

Comment (ehanson8) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:176]: I'm fine with whatever so long as it's not abbreviated 🙂

**Review by matt-bernhardt — APPROVED** [review]
I'm in favor of Option 3 as well. I'm not sure what the proper way in GraphQL is for changing a field type, so I'll defer to @JPrevost or @jazairi on that question.

Thanks, everyone, for working through this problem space together. I greatly appreciate everyone's contributions to this process.

File: docs/adrs/0003-support-aggregations-on-publisher-name.md
Comment (jazairi) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md:176]: Oh, yes, good point about the name conflict. I'd be happy with `Publishers` as an alternative.

Comment (ghukill) [reply docs/adrs/0003-support-aggregations-on-publisher-name.md]: @ehanson8, @jazairi - field name updated from `PublicationInfo` to `Publishers` in this most commit: https://github.com/MITLibraries/transmogrifier/pull/132/commits/9bb06e567a9195c4cdea92c48d6693a17169cf27

**Review by ghukill — APPROVED** [review]
While I originally supported Option 2, the comment from @matt-bernhardt about losing the context of dates and locations with respect to the publisher name was a deal breaker for me.

After much discussion and a new option 3 added, I support this decision as well.

**Review by JPrevost — COMMENTED** [review]
👍🏻 (I can't approve as I opened the PR, but I support the decision to go with Option 3)

**Pull request merged**
